Autopilot/PathManager/Src/landingMode/landingModeStages.cpp

```diff
@@ -0,0 +1,416 @@
+
+    //load in sensor fusion data and telemtry data into input structure
+    Telemetry_PIGO_t telem_data = land_mode->getTelemetryData();
+    _input.telemetryData = &telem_data;
+
+    SFOutput_t sf_data;
+    #ifdef UNIT_TESTING
+        sf_data.track = 1;
+    #else
+        sf_data = land_mode->getSensorFusionData();
+    #endif
+    _input.sensorOutput = &sf_data;
+
+    //making sure landing points are only made once
+    if(!land_mode->getMadeLandingPoints()) {
+        //requires data structure that dhruv wants to use
+        _path = LandingTakeoffManager::createSlopeWaypoints(*_input.telemetryData, _input.sensorOutput->altitude);
+
+        //creating waypoints
+        _path_array[0] = _landing_path.initialize_waypoint(_path.intersectionPoint.longitude, _path.intersectionPoint.latitude, _path.intersectionPoint.altitude, PATH_FOLLOW);
+        _path_array[1] = _landing_path.initialize_waypoint(_path.aimingPoint.longitude, _path.aimingPoint.latitude, _path.aimingPoint.altitude, PATH_FOLLOW);
+        _path_array[2] = _landing_path.initialize_waypoint(_path.stoppingPoint.longitude, _path.stoppingPoint.latitude, _path.stoppingPoint.altitude, PATH_FOLLOW);
+        _current_location = _landing_path.initialize_waypoint(_input.sensorOutput->longitude, _input.sensorOutput->latitude, _input.sensorOutput->altitude, HOLD_WAYPOINT, 20); //fill in with sensor fusion data
+
+        //initializing flight path
+        _waypoint_status = _landing_path.initialize_flight_path(_path_array, 3, _current_location);
+
+        //set made madelandingPoints to true
+        land_mode->setMadeLandingPoints(true);
+    }
+
+    //translate sensor data to waypoint data in struct
+    _waypoint_input.latitude = _input.sensorOutput->latitude;
+    _waypoint_input.longitude = _input.sensorOutput->longitude;
+    _waypoint_input.altitude = _input.sensorOutput->altitude;
+    _waypoint_input.track = _input.sensorOutput->track;
+
+    //follow the landing waypoints
+    _waypoint_status = _landing_path.get_next_directions(_waypoint_input, &_waypoint_output);
+
+    //translate waypoint commands into landing output structure
+    _output = LandingTakeoffManager::translateWaypointCommands(_waypoint_output);
+
+    //calculating the difference in heading to detect if finished turning (2 differences possible)
+    double difference_in_heading_1 = _input.telemetryData->stoppingDirectionHeading - _input.sensorOutput->track;
+    double difference_in_heading_2 = _input.sensorOutput->track - _input.telemetryData->stoppingDirectionHeading;
+
+    //making sure both headings are positive
+    if(difference_in_heading_1 < 0) {difference_in_heading_1 += 360;}
+    if(difference_in_heading_2 < 0) {difference_in_heading_2 += 360;}
+
+    //if the smaller heading is less than 5 degrees, set stage to slope
+    if((difference_in_heading_1 < difference_in_heading_2 && fabs(difference_in_heading_1) <= 5) || (fabs(difference_in_heading_2) <= 5)) {
```

Comment:
something's wrong here, in my last simulation testing PR, this should have been changed significantly. Not sure why those changes aren't coming up here, will look into it after reviewing this.

Reply:
yeah, this exists in the current devel repo, maybe rebase?

Reply:
oh fuck I didn't fetch upstream... shit this will be a nightmare to rebase...